src/field/prime/mod.rs

```diff
@@ -0,0 +1,259 @@
+//! This module contains abstractions specifically for fields of prime order.
+//! All the arithmetic operations are implemented for the [`PrimeField`] struct and the
+//! [`FiniteField`] trait is implemented. This module asserts at compile time that the order of the
+//! field is a prime number and allows for creation of generic prime order fields.
+
+use super::*;
+
+mod arithmetic;
+
+/// The two prime fields used in the Pluto `ronkathon` system.
+pub enum PlutoPrime {
+  /// The base field of curves used throughout the system.
+  Base = 101,
+
+  /// The scalar field for the curve over the [`PrimeField`] used in the polynomial commitment
+  /// scheme.
+  Scalar = 17,
+}
+
+/// The [`PrimeField`] struct represents elements of a field with prime order. The field is defined
+/// by a prime number `P`, and the elements are integers modulo `P`.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Default)]
+pub struct PrimeField<const P: usize> {
+  value: usize,
+}
+
+impl<const P: usize> PrimeField<P> {
+  /// Creates a new element of the [`PrimeField`] and will automatically compute the modulus and
+  /// return a congruent element between 0 and `P`. Given the `const fn is_prime`, a program that
+  /// tries to compile for a non-prime `P` will fail at compile time.
+  pub const fn new(value: usize) -> Self {
+    is_prime(P);
+    Self { value: value % P }
+  }
+}
+
+impl<const P: usize> const FiniteField for PrimeField<P> {
+  const GENERATOR: Self = if P == 2 { Self::ONE } else { find_generator::<P>() };
+  const NEG_ONE: Self = Self { value: P - 1 };
+  const ONE: Self = Self { value: 1 };
+  const ORDER: usize = P;
+  const THREE: Self = Self { value: 3 };
+  const TWO: Self = Self { value: 2 };
+  const ZERO: Self = Self { value: 0 };
+
+  fn inverse(&self) -> Option<Self> {
+    if self.value == 0 {
+      return None;
+    }
+    let exponent = Self::ORDER - 2;
+    let mut result = Self::ONE;
+    let mut base = *self;
+    let mut power = exponent;
+
+    while power > 0 {
+      if power & 1 == 1 {
+        result *= base;
+      }
+      base = base * base;
+      power >>= 1;
+    }
+    Some(result)
+  }
+
+  fn pow(self, power: usize) -> Self {
+    if power == 0 {
+      Self::ONE
+    } else if power == 1 {
+      self
+    } else if power % 2 == 0 {
+      Self::new((self.pow(power / 2).value * self.pow(power / 2).value) % Self::ORDER)
+    } else {
+      Self::new((self.pow(power / 2).value * self.pow(power / 2).value * self.value) % Self::ORDER)
+    }
+  }
+}
+
+const fn is_prime(n: usize) {
+  let mut i = 2;
+  while i * i <= n {
+    if n % i == 0 {
+      panic!("input is not a prime number");
+    }
+    i += 1;
+  }
+}
+
+const fn find_generator<const P: usize>() -> PrimeField<P> {
```

Comment:
some comments or a wiki link on what this algorithm is call or what the intuition here is would be sweet!